process/smartContract/builtInFunctions/keyValueStorage.go

```diff
@@ -13,17 +13,45 @@ import (
 
 var _ process.BuiltinFunction = (*saveKeyValueStorage)(nil)
 
+const numInitCharactersInProtectedKeys = 4
+
 type saveKeyValueStorage struct {
-	gasConfig   BaseOperationCost
-	funcGasCost uint64
+	gasConfig      BaseOperationCost
+	funcGasCost    uint64
+	mapInvalidKeys map[string]struct{}
 }
 
 // NewSaveKeyValueStorageFunc returns the key-value
 func NewSaveKeyValueStorageFunc(
 	gasConfig BaseOperationCost,
 	funcGasCost uint64,
-) *saveKeyValueStorage {
-	return &saveKeyValueStorage{gasConfig: gasConfig, funcGasCost: funcGasCost}
+	mapInvalidKeys map[string]struct{},
+) (*saveKeyValueStorage, error) {
+	if mapInvalidKeys == nil {
+		return nil, process.ErrNilInvalidKeysMap
+	}
+
+	s := &saveKeyValueStorage{
+		gasConfig:   gasConfig,
+		funcGasCost: funcGasCost,
+	}
+	s.mapInvalidKeys = make(map[string]struct{}, len(mapInvalidKeys))
+	for key := range mapInvalidKeys {
+		trimmedKey := key[:numInitCharactersInProtectedKeys]
+		s.mapInvalidKeys[trimmedKey] = struct{}{}
+	}
+
+	return s, nil
+}
+
+func (k *saveKeyValueStorage) isAllowedToSaveUnderKey(key []byte) bool {
```

Comment:
:+1:

Reply:
This check should be used in two more places: as a blockchain hook, for Arwen, to stop SmartContracts for overwriting protected keys, and in `processSCOutputAccounts()` as a final check, before individual storage updates are saved to the DataTrie.

Reply:
Or directly in the `DataTrieTracker()`, to prevent us from accidentally forgetting to check `isAllowedToSaveUnderKey()`.

Reply:
good catch. I will add at processSCOutput I think.

Reply:
I will skip those keys which are protected for these writes, maybe with a warning ? shouldn't we protect it at arwen level - write key value ?

Reply:
done